daemon/cmd/kube_proxy_replacement.go

```diff
@@ -228,33 +228,71 @@ func initKubeProxyReplacementOptions() (strict bool) {
 		probe.HaveIPv6Support()
 
 		option.Config.EnableHostServicesPeer = true
-		if option.Config.EnableIPv4 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET4_GETPEERNAME) != nil ||
-			option.Config.EnableIPv6 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET6_GETPEERNAME) != nil {
-			option.Config.EnableHostServicesPeer = false
+		if option.Config.EnableIPv4 {
+			if err := bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET4_GETPEERNAME); err != nil {
+				log.WithError(err)
+				option.Config.EnableHostServicesPeer = false
+			}
+		}
+		if option.Config.EnableIPv6 {
+			if err := bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET6_GETPEERNAME); err != nil {
+				log.WithError(err)
+				option.Config.EnableHostServicesPeer = false
+			}
 		}
 
-		if option.Config.EnableHostServicesTCP &&
-			(option.Config.EnableIPv4 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET4_CONNECT) != nil ||
-				option.Config.EnableIPv6 && bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET6_CONNECT) != nil) {
-			msg := "BPF host reachable services for TCP needs kernel 4.17.0 or newer."
-			if strict {
-				log.Fatal(msg)
-			} else {
-				option.Config.EnableHostServicesTCP = false
-				log.Warn(msg + " Disabling the feature.")
+		if option.Config.EnableHostServicesTCP {
+			var disableHostServicesTCP bool
+			if option.Config.EnableIPv4 {
+				if err := bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET4_CONNECT); err != nil {
+					log.WithError(err)
+					disableHostServicesTCP = true
+				}
+			}
+			if option.Config.EnableIPv6 {
+				if err := bpf.TestDummyProg(bpf.ProgTypeCgroupSockAddr, bpf.BPF_CGROUP_INET6_CONNECT); err != nil {
+					log.WithError(err)
+					disableHostServicesTCP = true
+				}
+			}
+
+			if disableHostServicesTCP {
+				msg := "BPF host reachable services for TCP needs kernel 4.17.0 or newer."
+				if strict {
+					log.Fatal(msg)
+				} else {
+					option.Config.EnableHostServicesTCP = false
+					log.Warn(msg + " Disabling the feature.")
+				}
 			}
```

Comment:
Maybe this could be refactored to use a function such as `probeCgroupSupport(tcp, ipv6 bool)`. We should then only call:
```go
if option.Config.EnableHostServicesTCP && option.Config.EnableIPv4 {
    probeCgroupSupport(true, false)
}
if option.Config.EnableHostServicesTCP && option.Config.EnableIPv6 {
    probeCgroupSupport(true, true)
}
if option.Config.EnableHostServicesUDP && option.Config.EnableIPv4 {
    probeCgroupSupport(false, false)
}
if option.Config.EnableHostServicesUDP && option.Config.EnableIPv6 {
    probeCgroupSupport(false, true)
}
```
The logic would also be the same as the original code as we would fatal as soon as we find something missing.

Reply:
I was worried that the code had blown up a bit. I actually added two functions for refactoring now to avoid even more `if {} else {}` statements. Hope thats fine. The logic should be the same as the original code now too.